`context.getScope()` returns an wrong scope when `blockBindings: true`.

I'm really surprised.

----

```
env:
    es6: true
ecmaFeatures:
    modules: true
rules:
    no-invalid-this: 2
```

```
> eslint -v
v1.0.0
> echo "function foo() { this; }" | eslint --stdin
>
```

No errors.
I expected an error, "Unexpected `this`."

Of course, this simple case exists in [its unit test](https://github.com/eslint/eslint/blob/master/tests/lib/rules/no-invalid-this.js#L108).
This problem happens only when `--env es6` was specified.

----

```
ecmaFeatures:
    modules: true
rules:
    no-invalid-this: 2
```

```
> eslint -v
v1.0.0
> echo "function foo() { this; }" | eslint --stdin

<text>
  1:19  error  Unexpected `this`  no-invalid-this

✖ 1 problem (1 error, 0 warnings)

>
```

It's OK if there is not `es6: true`.

----

If `--env es6` is specified, [here](https://github.com/eslint/eslint/blob/master/lib/rules/no-invalid-this.js#L297), `context.getScope()` returns the global scope. But this is always on `FunctionDeclaration` or `FunctionExpression`, so `context.getScope()` should return a function scope.

And the global scope is not strict mode, so this rule ignores the `this` keyword....

--- a/lib/eslint.js
+++ b/lib/eslint.js
@@ -1041,7 +1041,7 @@ module.exports = (function() {
             // if current node introduces a scope, add it to the list
             var current = controller.current();
             if (currentConfig.ecmaFeatures.blockBindings) {
-                if (["BlockStatement", "SwitchStatement", "ArrowFunctionExpression"].indexOf(current.type) >= 0) {
+                if (["BlockStatement", "SwitchStatement", "CatchClause", "FunctionDeclaration", "FunctionExpression", "ArrowFunctionExpression"].indexOf(current.type) >= 0) {
                     parents.push(current);
                 }
             } else {

--- a/tests/lib/eslint.js
+++ b/tests/lib/eslint.js
@@ -946,6 +946,45 @@ describe("eslint", function() {
 
             eslint.verify("if (true) { let x = 1 }", config, filename, true);
         });
+
+        it("should retrieve the function scope correctly from within a FunctionDeclaration", function() {
+            var config = { rules: {}, ecmaFeatures: { blockBindings: true } };
+
+            eslint.reset();
+            eslint.on("FunctionDeclaration", function() {
+                var scope = eslint.getScope();
+                assert.equal(scope.type, "function");
+                assert.equal(scope.block.type, "FunctionDeclaration");
+            });
+
+            eslint.verify("function foo() {}", config, filename, true);
+        });
+
+        it("should retrieve the function scope correctly from within a FunctionExpression", function() {
+            var config = { rules: {}, ecmaFeatures: { blockBindings: true } };
+
+            eslint.reset();
+            eslint.on("FunctionExpression", function() {
+                var scope = eslint.getScope();
+                assert.equal(scope.type, "function");
+                assert.equal(scope.block.type, "FunctionExpression");
+            });
+
+            eslint.verify("(function foo() {})();", config, filename, true);
+        });
+
+        it("should retrieve the catch scope correctly from within a CatchClause", function() {
+            var config = { rules: {}, ecmaFeatures: { blockBindings: true } };
+
+            eslint.reset();
+            eslint.on("CatchClause", function() {
+                var scope = eslint.getScope();
+                assert.equal(scope.type, "catch");
+                assert.equal(scope.block.type, "CatchClause");
+            });
+
+            eslint.verify("try {} catch (err) {}", config, filename, true);
+        });
     });
 
     describe("marking variables as used", function() {

--- a/tests/lib/rules/no-invalid-this.js
+++ b/tests/lib/rules/no-invalid-this.js
@@ -484,6 +484,15 @@ var patterns = [
         errors: errors,
         valid: [NORMAL, USE_STRICT, MODULES],
         invalid: []
+    },
+
+    // https://github.com/eslint/eslint/issues/3254
+    {
+        code: "function foo() { console.log(this); z(x => console.log(x, this)); }",
+        ecmaFeatures: {arrowFunctions: true, blockBindings: true},
+        errors: errors,
+        valid: [NORMAL],
+        invalid: [USE_STRICT, MODULES]
     }
 ];
 

